src/properties/degrees.c

```diff
@@ -517,3 +517,154 @@ igraph_error_t igraph_sort_vertex_ids_by_degree(const igraph_t *graph,
     }
     return IGRAPH_SUCCESS;
 }
+
+/**
+ * \function igraph_construct_jdm
+ * \brief Constructs a joint degree matrix.
+ *
+ * \experimental
+ *
+ * The joint degree matrix of a graph contains the number of edges, or the sum of the weights, between vertices of
+ * degree i and degree j for every (i, j). The function populates a joint degree matrix and works on directed
+ * and undirected graphs. Self loops are only counted once. If either \p dout or \p din are &lt;= 0, the JDM
+ * will be resized automatically in the dimensions of the max degree of the vertices in the graph.
+ *
+ * \param graph A pointer to an initialized graph object.
+ * \param jdm A pointer to an initialized integer matrix that will be resized. The values will be written here.
+ * \param dout An integer for the dimension of the matrix. If \p dout and \p din are greater than 0, the matrix will be
+ *        resized do dimensions (\p dout, \p din).
+ * \param din An integer for the dimension of the matrix. If \p dout and \p din are greater than 0, the matrix will be
+ *        resized do dimensions (\p dout, \p din).
+ * \param weights An initialized integer vector containing the weights of the edges. NULL if the JDM should sum
+ *        the number of edges rather than sum up the weights.
+ * \return Error code.
+ *
+ * Time complexity: O(E),
+ * where E is the number of edges in input graph.
+ */
+
+igraph_error_t igraph_construct_jdm(const igraph_t* graph,
+                                    igraph_matrix_int_t* jdm,
+                                    igraph_integer_t dout,
+                                    igraph_integer_t din,
+                                    const igraph_vector_int_t* weights) {
```

Comment:
I am just noticing one more issue that we need to fix.  We should operate with real-values weights, which implies that we should also use real-values matrices.

Please refactor to use `igraph_matrix_t` for the JDM and `igraph_vector_t` for `weights` instead of the current `igraph_matrix_int_t` and `igraph_vector_int_t`.